Status page certificate expiry

tarun7singh opened the pull request:
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]: 
- [x] I have read and understand the pull request rules.

# Description

Fixes #2451 
Fixes #3456
Fixes #2522 

## Type of change

- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I ran ESLint and other linters for modified files
- [X] I have performed a self-review of my own code and tested it
- [X] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [X] My changes generate no new warnings

## Screenshots (if any)

![Screenshot 2023-07-04 at 7 17 16 PM](https://github.com/louislam/uptime-kuma/assets/12068093/b5a57891-edb1-42fa-9f6d-5af86f12ebe3)

tarun7singh commented:
@louislam  This pull request is ready fo review, let me know if we need any style changes on how expiry is shown on the status page

CommanderStorm commented on server/model/status_page.js:
```diff
@@ -90,6 +90,9 @@ class StatusPage extends BeanModel {
      * @param {StatusPage} statusPage
      */
     static async getStatusPageData(statusPage) {
+
```
```suggestion
```

CommanderStorm commented on src/pages/StatusPage.vue:
```diff
@@ -54,6 +54,12 @@
+                <!-- Show certificate expiry -->
+                <div class="my-3 form-check form-switch">
+                    <input id="show-certificate-expiry" v-model="config.showCertificateExpiry" class="form-check-input" type="checkbox">
+                    <label class="form-check-label" for="show-certificate-expiry">{{ $t("Show Certificate Expiry") }}</label>
```
Could you add this key `Show Certificate Expiry` to `en.json` (and any languages for which you may be a native speaker)?
Otherwise, translators won't be able to translate this. ^^

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -154,6 +159,19 @@ export default {
+         */
+        formatExpiry(monitor) {
+            if (monitor?.element?.validCert) {
+                return monitor.element.certExpiryDaysRemaining + " days";
```
Could you make `days` translatable?

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -154,6 +159,19 @@ export default {
+            if (monitor?.element?.validCert) {
+                return monitor.element.certExpiryDaysRemaining + " days";
+            } else {
+                return monitor.element.certExpiryDaysRemaining;
```
If you rely on the server-generated message like this, `No/Bad Cert` will not be able to be translated. Is this an intentional decision?
(Could this be refactored to the frontend?)

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -154,6 +159,19 @@ export default {
+         * @param {Object} monitor Monitor to show expiry for
+         * @returns {string}
+         */
+        formatExpiry(monitor) {
```
Could `formattedCertExpiryMessage` be a better name?

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -60,6 +60,7 @@
                                                     @click="$refs.monitorSettingDialog.show(group, monitor)"
                                                 />
                                             </span>
+                                            <p v-if="showCertificateExpiry" class="item-name"> Expiry: {{ formatExpiry(monitor) }} </p>
```
Disclaimer: I am not a designer => feel free to overrule
From a design perspective, I don't like how this looks like a regular title (a user would not be able to pick up, that this is an expiry notice).

What do you think: Could we move this to a subtle or tag and still keep the current design?

tarun7singh replied:
Should i try showing it as pill like the uptime percetage, with colors for showing how soon it will be expiring?
My reason for avoiding that was to not clutter the ui.

CommanderStorm replied:
Yes, a pill is what I meant with tag. This would be one option.

tarun7singh replied:
1. Expiry inside the pill
![Screenshot 2023-07-06 at 1 36 50 PM](https://github.com/louislam/uptime-kuma/assets/12068093/36b90afa-0459-4f93-a62d-d795d2997247)

2. Expiry outside the pill
![Screenshot 2023-07-06 at 1 35 46 PM](https://github.com/louislam/uptime-kuma/assets/12068093/50888fa0-3b6e-468d-838c-40d00baca845)

Color is green for expiry of more than 7 days
Color is red for everything else

tarun7singh replied:
@louislam Do you have any preference out of the above two designs?

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -60,6 +60,7 @@
                                                     @click="$refs.monitorSettingDialog.show(group, monitor)"
                                                 />
                                             </span>
+                                            <p v-if="showCertificateExpiry" class="item-name"> Expiry: {{ formatExpiry(monitor) }} </p>
```
`Expiry` won't be able to be translated. Please add this as a translation key

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -154,6 +159,33 @@ export default {
+         * @returns {string}
+         */
+        certExpiryColor(monitor) {
+            if (monitor?.element?.validCert) {
+                return monitor.element.certExpiryDaysRemaining > 7 ? "primary" : "danger";
+            }
+            return "danger";
```
I think the following is more readable (unshure if nullisch coalessing is needed as well).
```suggestion
            if (monitor?.element?.validCert && monitor.element.certExpiryDaysRemaining > 7) {
                return "primary";
            }
            return "danger";
```

tarun7singh replied:
Fixed

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -154,6 +159,33 @@ export default {
+         * @returns {string}
+         */
+        formattedCertExpiryMessage(monitor) {
+            if (monitor?.element?.validCert && monitor?.element?.certExpiryDaysRemaining) {
```
I have not checked this: 
is `monitor?.element?.certExpiryDaysRemaining` necessary in this statement?

tarun7singh replied:
I'm doing this to avoid a scenario where certExpiryDaysRemaining are still loading(happens when you first turn the toggle and before saving), if not done like this it will show "Expiry  Day"

CommanderStorm replied:
Would `unknown` be a better option in this case?

tarun7singh replied:
Yes, that sounds better to me as well. I'm creating a fix for this and there was another bug that found adding fix for that as well.

CommanderStorm reviewed (approved):
My concerns have been addressed. (if you mark them as resolved, they are not shown any more ^^)

I (not a desiger/bdfl => this requires Louis vote) am unsure if the Pills' design fits, but I think this has a solid chance and looks close enough.

chakflying commented:
Expiry should not be shown if monitor type is not HTTPS.

louislam commented:
![image](https://github.com/louislam/uptime-kuma/assets/1336778/f4af43cb-d0fd-4c57-8d8e-8f7f7c5297fb)

I might prefer to put in front of the tag list. Like this.

tarun7singh commented:
I have incorporated tags design as suggested by @louislam and added check for monitor type (@chakflying ).
![Screenshot 2023-07-17 at 1 53 47 AM](https://github.com/louislam/uptime-kuma/assets/12068093/e08138bd-73a0-4d53-aede-4ec086a55040)

tarun7singh commented:
This is ready to be merged @CommanderStorm  & @louislam

CommanderStorm commented:
Could you add `Resolves #3456` to the description?

maurerle commented:
And also
resolves #2522

tarun7singh commented:
@maurerle 
I can add #2522 to the description but should we create a new issue for the sortable part?

CommanderStorm commented:
> should we create a new issue for the sortable part?

I think this is included in https://github.com/louislam/uptime-kuma/issues/1610 => no need to create another issue
